Chapter5/event_manager.erl

```diff
@@ -0,0 +1,89 @@
+-module(event_manager).
+-export([start/2, stop/1]).
+-export([add_handler/3, delete_handler/2, get_data/2, send_event/2, swap_handlers/4]).
+-export([init/1]).
+
+start(Name, HandlerList) ->
+	register(Name, spawn(event_manager, init, [HandlerList])),
+	ok.
+
+init(HandlerList) -> loop( initialize(HandlerList) ).
+
+initialize([]) -> [];
+initialize([{Handler, InitData}|Rest]) ->
+	[ { Handler, Handler:init(InitData) } | initialize(Rest) ].
+
+
+stop(Name) ->
+	Name ! {stop, self()},
+	receive
+		{reply, Reply} -> Reply
+	end.
+
+terminate([]) -> [];
+terminate([{Handler, Data}|Rest]) ->
+	[ { Handler, Handler:terminate(Data) } | terminate(Rest) ] .
+
+
+
+%%Exercise 5-3: Swapping Handlers (Page 162)
+swap_handlers(Name, OldHandler, NewHandler, InitData) ->
+	case add_handler(Name, NewHandler, InitData) of
```

Comment:
Ahh I didn't think about duplicating events across two handlers as a problem but I could see how it could be, updated the code to process the swap inside the loop so any events will be paused till swap is complete